include/Spectra/Util/SearchSpace.h

```diff
@@ -0,0 +1,80 @@
+// Copyright (C) 2016-2020 Yixuan Qiu <yixuan.qiu@cos.name>
+//
+// This Source Code Form is subject to the terms of the Mozilla
+// Public License v. 2.0. If a copy of the MPL was not distributed
+// with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef SPECTRA_SEARCH_SPACE_H
+#define SPECTRA_SEARCH_SPACE_H
+
+#include <Eigen/Dense>
+#include "RitzPairs.h"
+
+namespace Spectra {
+template <typename Scalar>
+class SearchSpace
+{
+private:
+    using Index = Eigen::Index;
+    using Matrix = Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>;
+    using Vector = Eigen::Matrix<Scalar, Eigen::Dynamic, 1>;
+    using Array = Eigen::Array<Scalar, Eigen::Dynamic, 1>;
+    using BoolArray = Eigen::Array<bool, Eigen::Dynamic, 1>;
+    using MapConstVec = Eigen::Map<const Vector>;
+
+public:
+    SearchSpace() = default;
+
+    void size() const
+    {
+        return basis_vectors_.cols();
+    }
+
+    template <typename OpType>
+    void update_operator_basis_product(OpType &op)
+    {
+        Index nvec = basis_vectors_.cols() - op_basis_product_.cols();
+        op_basis_product_.conservativeResize(Eigen::NoChange, basis_vectors_.cols());
+        op_basis_product_.rightCols(nvec) = op * basis_vectors_.rightCols(nvec);
+    }
+
+    template <typename OpType>
+    void full_update(OpType &op)
+    {
+        op_basis_product_ = op * basis_vectors_;
+    }
+
+    void restart(RitzPairs<Scalar> &ritz_pairs, Index size)
+    {
+        basis_vectors_ = ritz_pairs.Vectors().leftCols(size);
+        op_basis_product_ = op_basis_product_ * ritz_pairs.SmallRitzVectors().leftCols(size);
+    }
+
+    void append_new_vectors_to_basis(Matrix &new_vect)
+    {
+        Index num_update = new_vect.cols();
+        basis_vectors_.conservativeResize(Eigen::NoChange, basis_vectors_.cols() + num_update);
+        basis_vectors_.rightCols(new_vect.cols()) = new_vect;
+    }
+
+    void extend_basis(Matrix &new_vect)
+    {
+        Index num_update = new_vect.cols();
+        append_new_vectors_to_basis(new_vect);
+        //basis_vectors_ = orthogonalize(basis_vectors_, num_update);
+
+        //TODO orthogonalize
+    }
+
+    const Matrix &BasisVectors() const { return basis_vectors_; }
+    Matrix &BasisVectors() { return basis_vectors_; }
```

Comment:
Do you need this ? That might my total ignorance in C++ speaking there ;)

Reply:
yes we need a const accessor to get them out and a non const accessort to put in an initial guess